stdlib/public/SDK/Foundation/DateInterval.swift

```diff
@@ -156,10 +156,10 @@ public struct DateInterval : ReferenceConvertible, Comparable, Hashable, Codable
     }
     
     public var hashValue: Int {
-        var buf: (UInt, UInt) = (UInt(start.timeIntervalSinceReferenceDate), UInt(end.timeIntervalSinceReferenceDate))
+        var buf: (TimeInterval, TimeInterval) = (start.timeIntervalSinceReferenceDate, end.timeIntervalSinceReferenceDate)
         return withUnsafeMutablePointer(to: &buf) {
-            $0.withMemoryRebound(to: UInt8.self, capacity: 2 * MemoryLayout<UInt>.size / MemoryLayout<UInt8>.size) {
-                return Int(bitPattern: CFHashBytes($0, CFIndex(MemoryLayout<UInt>.size * 2)))
+            $0.withMemoryRebound(to: UInt8.self, capacity: 2 * MemoryLayout<TimeInterval>.size / MemoryLayout<UInt8>.size) {
```

Comment:
I don't believe this is a sound usage of `withMemoryRebound(to:capacity:)` (and neither was original code's usage) – the method requires that you rebind to a type that has the same size and alignment as the current `Pointee` type (e.g `UInt` to `Int`). The reasoning for this is because it uses the capacity you pass in order to bind the memory back to the original type, which in this case would result in too many instances being re-bound back to `TimeInterval`.

One alternative is to use `withUnsafeBytes(of:)`, for example (untested):

```swift
  return withUnsafeBytes(of: &buf) { buffer in
    Int(bitPattern: CFHashBytes(Array(buffer), buffer.count))
  }
```

Reply:
Array would not be ideal since that will deallocate after passing in, I would suggest to either bit cast since we are going directly to C or use , and also the internal function CFHashBytes takes a `void *` not a `const void *` because of "historical reasons"... however it does not mutate the pointer in any way so it is safe to cast to a mutable type.

Reply:
this would be reasonable:
```
return withUnsafeMutableBytes(of: &buf) {
    return CFHashBytes($0.baseAddress?.assumingMemoryBound(to: UInt8.self), $0.count)
}

```

Reply:
The array-to-pointer conversion should keep the `Array` valid for the duration of the call to `CFHashBytes`, so assuming `CFHashBytes` doesn't escape the pointer, there should be no problem there.

And ah, okay, if `CFHashBytes` requires a mutable pointer, you could do:

```swift
  return withUnsafeBytes(of: &buf) { buffer in
    var array = Array(buffer)
    return Int(bitPattern: CFHashBytes(&array, buffer.count))
  }
```

Reply:
So if I understand correctly:
* I should change it to:
```
return withUnsafeMutableBytes(of: &buf) {
    return CFHashBytes($0.baseAddress?.assumingMemoryBound(to: UInt8.self), $0.count)
}
```
* Change https://github.com/apple/swift-corelibs-foundation/blob/master/Foundation/DateInterval.swift#L153 also.
* Add unit test.

Is this correct?

Reply:
yes that is the most efficient way to correct the failure at hand while still being correct in a memory binding from my understanding on the binding side. 

ideally we would like to avoid allocations while hashing

Reply:
Ok, I'll try to do these changes tomorrow. Can somebody point me where should I add those tests?


I was only be able to found `Date` tests.

https://github.com/apple/swift-corelibs-foundation/blob/master/TestFoundation/TestDate.swift

Are there any `DateInterval` tests, am I missing something?

Reply:
https://github.com/apple/swift/blob/master/test/stdlib/TestCalendar.swift has a few `DateInterval` related test cases.

Reply:
https://github.com/apple/swift/blob/master/test/stdlib/TestDateInterval.swift, too.

Reply:
I'm a bit confused. I'm not sure why there are tests in both apple/swift and apple/swift-corelibs-foundation. What determines does it go into one or the other?

Where should these tests live? I'm assuming that I should only call `hashValue` for some old dateinterval and make sure that it doesn't crash.

Reply:
On Darwin, Swift depends on platform Foundation. Tests for the overlay that target Darwin should go in apple/swift as a result.

On non-Darwin, swift-corelibs-foundation depends on Swift. Tests for swift-corelibs-foundation outside of Darwin should go in apple/swift-corelibs-foundation.